query-planner-js/src/buildQueryPlan.ts

```diff
@@ -134,13 +139,171 @@ export function buildQueryPlan(
   return {
     kind: 'QueryPlan',
     node: nodes.length
-      // if an operation is a mutation, we run the root fields in sequence,
-      // otherwise we run them in parallel
-      ? flatWrap(isMutation ? 'Sequence' : 'Parallel', nodes)
+      ? // if an operation is a mutation, we run the root fields in sequence,
+        // otherwise we run them in parallel
+        flatWrap(isMutation ? 'Sequence' : 'Parallel', nodes)
       : undefined,
   };
 }
 
+export function findDirectivesOnNode<
+  T extends { directives?: readonly DirectiveNode[] },
+>(node: T, directiveName: string) {
+  return (
+    node.directives?.filter(
+      (directive) => directive.name.value === directiveName,
+    ) ?? []
+  );
+}
+
+function collectInclusionConditions(document: DocumentNode) {
+  const { operation, fragmentMap } = getOperationAndFragments(document);
+
+  const conditionalTree = buildConditionalTree(
+    operation.selectionSet,
+    fragmentMap,
+  );
+
+  let allTopLevelsHaveConditional = true;
+  const inclusionConditions: {
+    skip: boolean | string | null;
+    include: boolean | string | null;
+  }[] = [];
+  walkConditionalTree(conditionalTree, (node) => {
+    if (node === null || (Array.isArray(node) && node[0] === null)) {
+      allTopLevelsHaveConditional = false;
```

Comment:
Couldn't this short circuit the walk instead of continuing to build up `inclusionConditions`? Alternatively, perhaps you could even change `buildConditionalTree` to return `null` as soon as it encounters a `null` condition, so you could avoid the rest of the build and walk altogether.

Reply:
Agreed! A `ConditionalTree` can no longer have a `null` node, and `buildConditionalTree` will short-circuit return `null` when possible. Good call, and this simplifies the walking too!

5479a038